designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+
+Once the `getFormatter()` method got this change, we can update the specification of custom formatters without breakage in the future to support streaming.
+
+#### § The other methods
+
+The following methods receive an `AsyncIterable<LintResult>` object as the first argument then return an `AsyncIterableIterator<LintResult>` object because those methods are sandwiched between `executeOnFiles()` and formatters.
+
+- `outputFixes()`
+- `getErrorResults()`
+
+<details>
+<summary>Example: Use `outputFixes()` and `getErrorResults()`.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+let results = eslint.executeOnFiles(patterns)
+// Update the files of the results if needed
+// (This must be done before `--quiet`)
+if (process.argv.includes("--fix")) {
+    results = ESLint.outputFixes(results)
+}
+// Filter the results if needed
+if (process.argv.includes("--quiet")) {
+    results = ESLint.getErrorResults(results)
+}
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
+
+The following methods return `Promise` which gets fulfilled with each result in order to support plugins/configs that are ES modules without breakage in the future.
+
+- `getConfigForFile()`
+- `isPathIgnored()`
+
+The following method is as-is because those don't touch both file system and module system.
+
+- `getRules()`
+
+The following methods are removed because those don't fit the current API.
+
+- `addPlugin()` ... This method has caused to confuse people. We have introduced this method to add plugin implementations and expected people to use this method to test plugins, but people have thought that this method loads a new plugin for the following linting. And this method is only one that mutates the state of `CLIEngine` objects (except cache). This RFC moves this functionality to the constructor options.
+- `resolveFileGlobPatterns()` ... ESLint doesn't use this logic since `v6.0.0`, but it has stayed there for backward compatibility. Once [RFC 20](https://github.com/eslint/rfcs/tree/master/designs/2019-additional-lint-targets) is implemented, what ESLint iterates and what the glob of this method iterates will be different, then it will confuse users. This is good timing to remove the legacy.
+
+#### § A new `compareResultsByFilePath` static method
+
+This method receives two lint results then returns `+1`, `-1`, or `0`. This method is intended to use in order to sort results.
+
+<details>
+<summary>A rough sketch of the `compareResultsByFilePath()` method.</summary>
+
+```js
+class ESLint {
+  static compareResultsByFilePath(a, b) {
+    if (a.filePath < b.filePath) {
+      return -1
+    }
+    if (a.filePath > b.filePath) {
+      return 1
+    }
+    return 0
+  }
+}
+```
+
+</details>
+
+### ■ Deprecate `CLIEngine` class
+
+This RFC soft-deprecates `CLIEngine` class.
```

Comment:
May I suggest adding something similar to:

```markdown
- Adds `disallowWorkerThreads` option to plugins. Defaults to `false`.
```

Like:

```markdown
- Adds `requiresAsyncUse` option to plugins. Defaults to `false`.
```

That way any plugin which wants to move to fully use the new async features can do so without risking incompatibilities with the sync flow.

If that property is set on a plugin, then the sync flow of `CLIEngine` should error when encountering it and tell the user about the incompatibility.

This would help the ecosystem in regards to the deprecation and the move towards an async flow.

Reply:
I'm not sure if there are incompatibility things about this. We don't have any asynchronous API to plugins. On the other hand, running in parallel can break plugins that use OS resources exclusively (e.g., cache file).

In the future, if we support asynchronous stuff for plugins (e.g., ES modules. It requires `import()` to load the plugins), `CLIEngine` fails to load those. I don't think that we need additional flags.